### ISSUE TITLE
how to do Grid and Rondom proposal boxes initialization?

### ISSUE THREAD

**Issue by liz6688** [PeizeSun/SparseR-CNN#71]
## ❓ How to do something using detectron2

Describe what you want to do, including:
1. what inputs you will provide, if any:
2. what outputs you are expecting:

## ❓ What does an API do and how to use it?
Please link to which API or documentation you're asking about from
https://detectron2.readthedocs.io/

For meaning of a config, please see
https://detectron2.readthedocs.io/modules/config.html#config-references

NOTE:

1. Only general answers are provided.
   If you want to ask about "why X did not work" for something you did, please use the
   [Unexpected behaviors](https://github.com/facebookresearch/detectron2/issues/new/choose) issue template.

2. About how to implement new models / new dataloader / new training logic, etc., check documentation first.

3. We do not answer machine learning / computer vision questions that are not specific to detectron2, such as how a model works, how to improve your training/make it converge, or what algorithm/methods can be used to achieve X.

**Comment by PeizeSun** [comment]
```python
        self.init_proposal_boxes = nn.Embedding(self.num_proposals, 4)
        if self.init_type == "img":
            nn.init.constant_(self.init_proposals.weight[:, :2], 0.5)
            nn.init.constant_(self.init_proposals.weight[:, 2:], 1.0)
        elif self.anchor_type == "center":
            nn.init.constant_(self.init_proposals.weight[:, :2], 0.5)
            nn.init.constant_(self.init_proposals.weight[:, 2:], 0.1)
        elif self.anchor_type == "grid":
            scale = math.sqrt(1. / self.num_queries)
            steps = int(math.sqrt(self.num_queries)) + 1
            x_range = torch.linspace(0., 1., steps)[:-1] + scale / 2
            y_range = torch.linspace(0., 1., steps)[:-1] + scale / 2
            y_range, x_range = torch.meshgrid(y_range, x_range)
            centers = torch.cat([x_range.reshape(-1, 1), y_range.reshape(-1, 1)], 1)
            nn.init.constant_(self.init_proposals.weight[:, 2:], scale)
            for i in range(self.num_queries):
                nn.init.constant_(self.init_proposals.weight[i, 0], centers[i, 0])
                nn.init.constant_(self.init_proposals.weight[i, 1], centers[i, 1])
        elif self.anchor_type == "random":
            pass
```

**Comment by PeizeSun** [comment]
@liz6688 If you find any bug when using different initializations, please remind me, thanks~

**Comment by liz6688** [comment]
Got it. Thanks.

**Comment by wzic** [comment]
Hi! Thanks for the code.

When I used the "Random" initialization to do the training, the program reported error "assert (boxes[:, 2:] >= boxes1[:, :2]).all()" for the "generalized_box_iou" function in util/box_ops.py. I suppose the error comes from the situation that the box at some time gets a negative width or height. May I know if the "Random" intialization is responsible for this phenomenon? How could we do the "Random" initialization safety and avoid this problem? Are there any extra constraints needed?

Thank you in advance.

**Comment by luzibuye** [comment]
> How could we do the "Random" initialization safety and avoid this problem? Are there any extra constraints needed?

I followed the way in the above code to random initialize proposal boxes with mmdetection but the loss became very big and cannot train. Have you solved the "Random" initialization problem? :)
